### ISSUE TITLE
Sass files with font path variable

### ISSUE THREAD

**Issue by cbirdsong** [fontsource/fontsource#64]
Fontsource's .css files assume the font files are located at ./files/, which dictates where you must store them on your production. Build processes like Webpack can take care of this, but I rarely use complicated build tools like that, as the complexity quickly outweighs the benefits for less Javascript-centric contexts like Wordpress themes or static site generators

An easy solution would be including nearly identical .scss files alongside the existing .css files. The only change would be adding a `!default` path variable at the top, like so:

```
$font-file-path: './files' !default;

/* montserrat-latin-400-normal*/
@font-face {
  font-family: 'Montserrat';
  font-style: normal;
  font-display: swap;
  font-weight: 400;
  src:
    local('Montserrat Regular'),
    local('Montserrat-Regular'), 
    url('#{$font-file-path}/montserrat-latin-400-normal.woff2') format('woff2'), /* Chrome 26+, Opera 23+, Firefox 39+ */
    url('#{$font-file-path}/montserrat-latin-400-normal.woff') format('woff'); /* Chrome 6+, Firefox 3.6+, IE 9+, Safari 5.1+ */
}
```

This would allow the path to be overridden when imported, adding a lot of flexibility in how Fontsource's fonts can be used without additional complexity.

Additionally, it's easy to imagine more Sass variables that would enable powerful case-by-case customization that builds on Fontsource's generally reasonable defaults. (A short list: setting a different value for font-display, disabling use of local() fonts, setting a unicode-range based on the content in a build.)

**Comment by ayuhito** [comment]
I'm all for making Fontsource as framework agnostic as possible and this shouldn't be too relatively difficult to implement. 

I don't believe this should cause any problems if we have all SCSS files in its own folder for the sake of organisation, correct? Other than changing up the relative file paths.

Could you possibly expand on how we would go about disabling the use of local fonts and setting unicode-ranges based on the content in a build?

**Comment by cbirdsong** [comment]
A functional-but-inelegant way* to write it would be:

```
$font-file-path: './files' !default;
$font-display: 'swap' !default;
$unicode-range: null !default;
$use-local-fonts: true !default;

/* montserrat-latin-400-normal*/
@font-face {
    font-family: 'Montserrat';
    font-style: normal;
    font-display: $font-display;
    font-weight: 400;

    @if $use-local-fonts {
        src:
    	    local('Montserrat Regular'), 
            local('Montserrat-Regular'), 
    	    url('#{$font-file-path}/montserrat-latin-400-normal.woff2') format('woff2'),
    	    url('#{$font-file-path}/montserrat-latin-400-normal.woff') format('woff'); 
    } @else {
	src:
    	    url('#{$font-file-path}/montserrat-latin-400-normal.woff2') format('woff2'),
    	    url('#{$font-file-path}/montserrat-latin-400-normal.woff') format('woff'); 
    }

    @if $unicode-range {
	unicode-range: $unicode-range;
    }
}
```
_*Right off the top of my head, so there might be a cleaner way to do it without repeating so much code._ 

When importing this file, I would do something like:

```
$font-file-path: '/wp-content/themes/whatever/fonts`;
$font-display: optional;
$use-local-fonts: false;

@import "montserrat-latin-400-normal";
```
which would output:
```
/* montserrat-latin-400-normal*/
@font-face {
  font-family: "Montserrat";
  font-style: normal;
  font-display: optional;
  font-weight: 400;
  src: url("./wp-content/themes/whatever/fonts/montserrat-latin-400-normal.woff2") format("woff2"), url("./wp-content/themes/whatever/fonts/montserrat-latin-400-normal.woff") format("woff");
}
```

---

`unicode-range` is a lot more complex and probably could be skipped for v1 of this idea, but being able to set the path would be super helpful when using it. A (relatively) simple use would be if you were importing a monospace font exclusively for displaying numbers. List out the numerals and maybe some punctuation, run a [subsetter](https://fonttools.readthedocs.io/en/latest/subset/index.html) on the font files, and then do your Sass compilation.

If you want to get _quite a bit_ more complicated, if you had a design that only used a font in the page titles your build tool could loops through all the pages, get all the characters out of the title, add all those together, and then write that into a single-variable Sass file before triggering the Sass compilation and subsetting the font files.

**Comment by ayuhito** [comment]
The spec seems simple enough. I'm totally fine with the implementation of overhauling how we approach SASS.

I've looked into a subsetter system in the past for Fontsource, but it's certainly an extremely complex undertaking that I'm not sure will be feasible for the simplicity this project is supposed to offer. I'm not crossing it out, but it will need a lot lot more planning and thought if it is even suitable.

If anyone is willing to create a PR for this, feel free to do so. Otherwise, I'll work on an implementation in the coming month after I finish working on adding variable font support.

Things to note when developing:
- Some fonts may not have local references in their CSS;
- Generate all files into a separate SCSS folder.

**Comment by ayuhito** [comment]
I'm no expert on SASS, but I can assume we can take inspiration from [`roboto-fontface`](https://github.com/choffmeister/roboto-fontface-bower/tree/master/css) and use mixins to simplify the build. Just dropping this as a note for future reference.

**Comment by ayuhito** [comment]
@cbirdsong, thanks for the issue. I've now released this as part of the [4.1.0 update](https://github.com/fontsource/fontsource/blob/master/CHANGELOG.md#41x). Feel free to create new issues that relate to this update if any bugs or ideas pop up!

**Comment by cbirdsong** [comment]
This is great! I’ll switch a project over to use these after the holidays and report back with any issues I encounter.

**Comment by davidjb** [comment]
Thank you @DecliningLotus - SCSS support works well.  

The only fiddliness I encountered was namespacing because each package defines its own `fontFace` mixin -- meaning that in my case where I'm building a CSS file which includes multiple typefaces, I just have to be aware of this overlap.  I'm using it like so https://github.com/jcu/cookbook/blob/master/src/scss/fonts.scss if that's of any interest to anyone -- creates a file that includes various weights and styles from multiple typefaces.

**Comment by ayuhito** [comment]
@davidjb, I had a concern on the namespacing being too generic and then completely forgot about that. ;-;

Ideally, wouldn't it be better for us to ship mixin variables with names like fontfaceOpenSans() or fontFaceRoboto() instead? That would making importing multiple typefaces a little less messy, cause less overlaps and be a bit more intuitive. 

In that case, we could just add those mixin variables and deprecate the generic ones for the next major update.

Is that a valid improvement in your eyes?

**Comment by cbirdsong** [comment]
@davidjb That was also awkward for me, but I figured out [the new Sass module system](https://css-tricks.com/introducing-sass-modules/) solves it pretty elegantly. (see below)

The more awkward thing for me was the directory. I ended up doing this, which still required a lot of repetition and hard-coded paths:

```
@use "@fontsource/open-sans/scss/mixins" as OpenSans;
@use "@fontsource/barlow-condensed/scss/mixins" as Barlow;
$themePath: "/wp-content/themes/a-cool-theme";

// Body
@include OpenSans.fontFace(
	$weight: 400,
	$style: normal,
	$fontDir: $themePath + "/node_modules/@fontsource/open-sans/files",
	$unicodeMap: latin
);
@include OpenSans.fontFace(
	$weight: 400,
	$style: italic,
	$fontDir: $themePath + "/node_modules/@fontsource/open-sans/files",
	$unicodeMap: latin
);

@include OpenSans.fontFace(
	$weight: 600,
	$style: normal,
	$fontDir: $themePath + "/node_modules/@fontsource/open-sans/files",
	$unicodeMap: latin
);
@include OpenSans.fontFace(
	$weight: 600,
	$style: italic,
	$fontDir: $themePath + "/node_modules/@fontsource/open-sans/files",
	$unicodeMap: latin
);


// Headings

@include Barlow.fontFace(
	$weight: 400,
	$fontDir: $themePath + "/node_modules/@fontsource/barlow-condensed/files",
	$unicodeMap: latin
);
```

This is still a big improvement over my previous method, which essentially was just writing the output of these mixins manually. However, some smart `!default` variables for the arguments in the mixin could really reduce the amount of code for each `@include`, since the directory, font-display and unicode range are likely to be the same project-wide.

For example, I just changed my local mixin files like so: 

```
$fontsourceDirLocation: "~" !default;
$fontsourceDisplay: swap !default;

@mixin fontFace(
  $fontName: "Open Sans",
  $fontId: "open-sans",
  $style: normal,
  $display: $fontsourceDisplay,
  $weight: 400,
  $fontDir: "#{$fontsourceDirLocation}@fontsource/#{$fontId}/files",
  $unicodeMap: (cyrillic-ext: (U+0460-052F, U+1C80-1C88, U+20B4, U+2DE0-2DFF, U+A640-A69F, U+FE2E-FE2F), cyrillic: (U+0400-045F, U+0490-0491, U+04B0-04B1, U+2116), greek-ext: (U+1F00-1FFF), greek: (U+0370-03FF), vietnamese: (U+0102-0103, U+0110-0111, U+0128-0129, U+0168-0169, U+01A0-01A1, U+01AF-01B0, U+1EA0-1EF9, U+20AB), latin-ext: (U+0100-024F, U+0259, U+1E00-1EFF, U+2020, U+20A0-20AB, U+20AD-20CF, U+2113, U+2C60-2C7F, U+A720-A7FF), latin: (U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD))
) {
  @each $subset, $unicodeRangeValues in $unicodeMap {
    // unmodified
  }
}
```

This means the font calls in my project can be as simple as:

```
@use "@fontsource/open-sans/scss/mixins" as OpenSans;
@use "@fontsource/barlow-condensed/scss/mixins" as Barlow;
$fontsourceDirLocation: "/wp-content/themes/a-cool-theme/node_modules/";

// Body
@include OpenSans.fontFace($weight: 400, $style: normal, $unicodeMap: latin);
@include OpenSans.fontFace($weight: 400, $style: italic, $unicodeMap: latin);

@include OpenSans.fontFace($weight: 600, $style: normal, $unicodeMap: latin);
@include OpenSans.fontFace($weight: 600, $style: italic, $unicodeMap: latin);

// Headings
@include Barlow.fontFace($weight: 600, $unicodeMap: latin);
```

Much cleaner! 

I'm not sure how easily this could be done for `$unicodeMap` right offhand, but I'm also not sure I'm actually using it correctly?

**Comment by cbirdsong** [comment]
@DecliningLotus I'm also really unsure how these are intended to work with variable fonts. Do you want me to open a separate issue?

**Comment by ayuhito** [comment]
@cbirdsong, I think this might be a matter of lack of documentation in this case.

For your issue regarding default variables, I would argue they already exist within the mixins.

For example, this:
```
@include OpenSans.fontFace(
	$weight: 400,
	$style: normal,
	$fontDir: $themePath + "/node_modules/@fontsource/open-sans/files",
	$unicodeMap: latin
);
```
can be simplified to this:
`@include OpenSans.fontFace($fontDir: $themePath + "/node_modules/@fontsource/open-sans/files");`

In this [mixin excerpt](https://github.com/fontsource/fontsource/blob/master/packages/open-sans/scss/mixins.scss), the `fontface($fontName: "Open Sans", ...) { ... }` values already being inputted are the default variables that don't need to be included. Any other repeated values should be variables created by the user.

I should also put a disclaimer that unicode-map should NOT be touched or included in your mixin input unless you know what you are doing. I only included the customisability because there may be extremely advanced niche cases where it makes sense to do that. That map is a bunch of key-value pairs that help generate the [duplicate CSS](https://github.com/fontsource/fontsource/blob/master/packages/open-sans/400.css) needed for the different unicode-range subsets.

I would love to make the documentation more in-depth with lots more examples and usecases, but our README.md is getting a bit long. I do want to move dedicated documentation over to the [search-directory](https://github.com/fontsource/search-directory), but that's going to be a little hard to find time for.

> I'm also really unsure how these are intended to work with variable fonts. Do you want me to open a separate issue?

Sure. The current workaround would be to use the fontFaceCustom() mixin and fill in the variables manually, which probably isn't the most efficient way of handling things. I've got a rough idea of introducing a new mixin for variable fonts so we should organise that in a separate issue.

**Comment by cbirdsong** [comment]
Definitely clear I’ve not used/understood "$unicodeMap` correctly. I will make another issue to keep discussion organized.

For defaults, relying on the mixin arguments is fine for things that are likely to change each time you call the mixin, but it seems really unlikely you'd change font-display, file path and character set for each individual typeface. That feels like an advanced feature and it's good the mixin allows for it, but it introduces a lot of repetition. 

I imagine the far more common scenario is someone deciding "we're using `font-display: optional;` and keeping our compiled files in `/dist/`", and those people have to repeat those instructions to each "@include fontFace` over and over again. Compare these two code blocks:

```
@use "@fontsource/open-sans/scss/mixins" as OpenSans;
@use "@fontsource/barlow-condensed/scss/mixins" as Barlow;

@include OpenSans.fontFace($fontDir: "/dist/node_modules/@fontsource/open-sans/files", $display: optional);
@include OpenSans.fontFace($fontDir: "/dist/node_modules/@fontsource/open-sans/files", $display: optional, $style: italic);

@include OpenSans.fontFace($fontDir: "/dist/node_modules/@fontsource/open-sans/files", $display: optional, $weight: 700);
@include OpenSans.fontFace($fontDir: "/dist/node_modules/@fontsource/open-sans/files",  $display: optional, $weight: 700,  $style: "italic");

@include Barlow.fontFace($fontDir: "/dist/node_modules/@fontsource/barlow-condensed/files", $display: optional, $weight: 600);
@include Barlow.fontFace($fontDir: "/dist/node_modules/@fontsource/barlow-condensed/files",  $display: optional, $weight: 600,  $style: "italic");
```
```
@use "@fontsource/open-sans/scss/mixins" as OpenSans;
@use "@fontsource/barlow-condensed/scss/mixins" as Barlow;

$fontsourceDir: "/dist/node_modules/@fontsource/"
$fontsourceDisplay: optional;

@include OpenSans.fontFace($weight: 400, $style: normal);
@include OpenSans.fontFace($weight: 400, $style: italic);

@include OpenSans.fontFace($weight: 700, $style: normal);
@include OpenSans.fontFace($weight: 700, $style: "italic");

@include Barlow.fontFace($weight: 600, $style: normal);
@include Barlow.fontFace($weight: 600,  $style: "italic");
```

The second way leads to much more maintainable code than the first. You can understand what typefaces are being loaded at a glance*, adding another weight is a simple copy/paste, and modifying the file path requires editing only a single line.

_*This is why I add the unnecessary `$weight: 400` and `$style: normal`._

**Comment by davidjb** [comment]
> @davidjb, I had a concern on the namespacing being too generic and then completely forgot about that. ;-;
> 
> Ideally, wouldn't it be better for us to ship mixin variables with names like fontfaceOpenSans() or fontFaceRoboto() instead? That would making importing multiple typefaces a little less messy, cause less overlaps and be a bit more intuitive.
> 
> In that case, we could just add those mixin variables and deprecate the generic ones for the next major update.
> 
> Is that a valid improvement in your eyes?

The setup at the moment with matching naming, at least for my setup, is pretty good.  The generic naming means my looping, mixin/include based approach is about as DRY as one can make it (since mixins can't be variables or referenced functionally).

At present, I'm building a Bootstrap 4-based theme being compiled by `node-sass` (so no Sass modules are available) but since Bootstrap 5 switches from libsass to Dart SASS, that will open up options like yours, @cbirdsong.

In short, the current implementation is great for me, @DecliningLotus and on balance, I think keeping the generic naming makes sense as it creates consistency across packages. Unless there's some other pressing reason, how you've got it is good. Thank you again!

**Comment by ayuhito** [comment]
@cbirdsong, I see. That definitely makes sense. Will work on it for the next update. Thank you for the explanation.

@davidjb, there seems to be a conflict on which is better at least. The problem with the consistency argument is that it can be a little conflicting and less intuitive when you are importing multiple typefaces.

We could go along with the earlier proposal where we ship fontFaceFontName() and fontFace(), without deprecating the latter, but I'm not sure if it'll cause confusion down the line. Alternatively, we could update the documentation and use @use ... as FontName, similar to @cbirdsong's examples, which takes away the need to ship the extra mixin namespace, whilst still accommodating your usecase for older Sass implementations.

I'm liking the latter option personally.

**Comment by cbirdsong** [comment]
`fontFaceFontName()` as a legacy option would work, but I will mull it over and see if I come up with any options that might be less awkward.

The documentation should probably prefer `@use` since [Node Sass/libsass has been deprecated](https://sass-lang.com/blog/libsass-is-deprecated), and if `fontFaceFontName()` is shipped for legacy use it should probably include an `@warn` message about that? 

As an aside, switching from Node Sass to Dart Sass is generally pretty easy! In my experience, you can usually just swap the `node-sass` package out for `sass` in your package.json.

**Comment by ayuhito** [comment]
> `fontFaceFontName()` as a legacy option would work, but I will mull it over and see if I come up with any options that might be less awkward.

Yeah, I feel that whole process is awkward. Committing to one format would be most appropriate to reduce confusion among the userbase and having clear documentation. I'll make a new issue to organise this update since I feel there is a lot of divergence to the original issue.

**Comment by ayuhito** [comment]
Closed in favour of #129 and #130.
